index.js

```diff
@@ -40,12 +40,47 @@ function isValidSchema (schema, name) {
   }
 }
 
-function mergeLocation (location, key) {
-  return {
-    schema: location.schema[key],
-    schemaId: location.schemaId,
-    jsonPointer: location.jsonPointer + '/' + key,
-    isValidated: location.isValidated
+class Location {
+  constructor (schema, schemaId, jsonPointer = '#', isValidated = false) {
+    this.schema = schema
+    this.schemaId = schemaId
+    this.jsonPointer = jsonPointer
+    this.isValidated = isValidated
+
+    this.isMerged = false
+    this.mergedSchemaId = null
+    this.mergedJsonPointer = null
```

Comment:
It is currently hard to follow why does it fix the issue and how it is fixed.
Would you mind to add some comment of step flow why it works? and how the usage of `mergedJsonPointer`?

Reply:
@climba03003 @Eomm  I will explain with an example. Sorry, it will not be simple because the case is not simple.

```js
{
  $id: 'root',
  type: 'object',
  properties: {
    property1: { type: 'string' },
    property2: {
      type: 'object',
      allOf: [
        {
          properties: {
            property21: { $ref: '#/properties/property1' }
          }
        },
        {
          properties: {
            property22: {
              anyOf: [
                { type: 'string' },
                { type: 'number' }
              ]
            }
          }
        }
      ]
    }
  }
}
```

1. When we meet property `allOf`, we create a merged schema and create a new unique id for it. 
https://github.com/fastify/fast-json-stringify/blob/e6b3c5fb6e1dc47f1c71560adf2ef3363bbf2b70/index.js#L458
If we meet anyOf/oneOf/if inside the merged schema, we create a reference using the merged schema id.
`property22` will have the next reference: `merged_12...223#/properties/property22`. To create this reference, we should know the merged schema id.

2. If somewhere inside the `allOf` schemas we meet a local reference like  `'#/properties/property1'` we have to resolve it. To do this, we should know the original schema reference, which is `root` in this case.

So that is why we need to keep two schemas ids when we build a serializer inside the `allOf` property.

Reply:
The `Location` class now has two methods for getting schema id.
1. `getSchemaId` - you should use it when you want to CREATE a new reference to some place in the schema. For example, for validation.
2. `getOriginSchemaId` - you should use it when you want to RESOLVE already existing references.

I added corresponding comments in the code.